pydoit.md

```diff
@@ -0,0 +1,138 @@
+    - HPC machine
+
+`pydoit` builds a database (in the folder of the `doit.py`) that tracks (via hashes of the *dependencies*) which *actions* need to be rerun. That means that each task has to be executed at least once, even if the *targets* already exist or are up-to-date.
+Problems arise, if due to some of the reasons above, the current machine is unable to do so. 
+
+One remedy is to _store_ a *target* in our version control system and tell `doit`: 
+*Trust me, the target in the version control is up-to-date.*
+
+### Proposed setup:
+~~~py
+    /data/dodo.py           # generates input data
+    /simulations/dodo.py    # turns input data into results
+    /images/dodo.py         # turns results into images
+    /paper/dodo.py          # turns images into a paper
+    /dodo.py       # optionally connect all tasks
+~~~
+
+The idea is to decouple the whole workflow into multiple `dodo` files. Someone, who just wants to build the `paper.pdf`, relies on up-to-date files in `/images/` and can simply run `doit` in the `/paper/` directory. 
+If one wants to execute the _full_ workflow, calling the root `doit` in `/`, will trigger the complete workflow.
+
+If you work with _absolute_ paths in your `dodo` files, such as ...
+~~~py
+# /simulation/dodo.py
+from pathlib import Path
+def task_simulation_1():
+    simulation_dir = Path(__file__).absolute().parent
+    script = str(simulation_dir / "simulation_1.py")
+    input = str(simulation_dir.parent / "data" / "input_1.dat")
+    image = str(simulation_dir.parent / "images" / "graph_1.pdf")
+    return {
+        "file_dep": [script, input],
+        "targets": [image],
+        "actions": [f"{python3 {script} {input}]
+    }
+~~~
+
+... and ...
+~~~py
+# /paper/dodo.py
+from pathlib import Path
+def task_simulation_1():
+    paper_dir = Path(__file__).absolute().parent
+    tex = str(paper_dir / "paper.tex")
+    pdf = str(paper_dir / "paper.pdf")
+    image = str(paper_dir.parent / "images" / "graph_1.pdf")
+    return {
+        "file_dep": [tex, image],
+        "targets": [pdf],
+        "actions": [f"{pdflatex {tex}]
+    }
+~~~
+... that makes it easy to connect the `image` target of the simulation task with the same `image` dependency of the paper task by just importing both into the same _root_ file
+
+~~~py
+# (root) /dodo.py
+from simulation.dodo import *
+from paper.dodo import *
+~~~
+
+
+Skipping huge simulations
+-------------------------
+
+If you just want to skip tasks (e.g. if they are too time consuming) you can define a flag and skip the task definition, if that flag is set. The simplest way would be to manually set the flag in the `dodo.py` file like
+
+~~~py
+import os
+def skip():
+    return True # set that to false manually
+
+def task_huge_simulation():
+    if skip():
+        return
+    return {...} # file_dep, actions, dependencies
+~~~
+
+It is advantageous to be able to change the flag without modifying the file, e.g. via an environment variable. 
```

Comment:
The `dodo.py` is not an automatic dependency of any task. You can modify it without causing doit to rerun up-to-date tasks.

Reply:
Sure! Still annoying regarding the diff.

Reply:
You're right, didn't think of that.